src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs

```diff
@@ -336,11 +359,75 @@ public void CopyChangedValues(TStructuralType original)
                     bool isDeltaType = TypedDelta.IsDeltaOfT(deltaNestedResource.GetType());
                     Contract.Assert(isDeltaType, nestedResourceName + "'s corresponding value should be Delta<T> type but is not.");
 
+                    Type nestedType = deltaNestedResource.StructuredType;
+                    Type originalType = originalNestedResource.GetType();
+
+                    if (deltaNestedResource.IsComplexType && !nestedType.IsAssignableFrom(originalType))
+                    {
+                        originalNestedResource = ReAssignComplexDerivedType(original, nestedResourceName, originalNestedResource, nestedType, originalType);
+                    }
+
                     deltaNestedResource.CopyChangedValues(originalNestedResource);
                 }
             }
         }
 
+        private dynamic ReAssignComplexDerivedType(TStructuralType original, string nestedResourceName, dynamic originalNestedResource, Type nestedType, Type originalType)
+        {
+            //As per OASIS discussion, changing a complex type from 1 derived type to another is allowed if both derived type have a common ancestor and the property
+            //is declared in terms of a common ancestor. The logic below checks for a common ancestor. Create a new object of the derived type in delta request.
+            //And copy the common properties.
+
+            Type complexBaseType = nestedType.BaseType;
```

Comment:
>complexBaseType [](http://example.com/codeflow?start=17&length=15)

complexBaseType doesn't help me understand if this is the base type of the new type or the original type.  Maybe call it newBaseType?